### ISSUE TITLE
Store image in local docker daemon when using pack build --publish

### ISSUE THREAD

**Issue by ipmb** [buildpacks/pack#1024]
### Description

If I use the `--publish` flag with `pack build`, the image is pushed to the remote repository as expected, but it is not available in the local docker daemon as it is without the `--publish` flag.

### Proposed solution

I expect the image to be available in the local Docker daemon _and_ be published to the remote repository. I want to run `pack build`, push the results to the remote repository, then execute tests against the image locally.

### Describe alternatives you've considered

Previously, I didn't use the `--publish` flag and pushed the image manually with docker, but now I want to use the `cache-image` feature which requires the `--publish` flag.

### Additional context
- [ ] This feature should be documented somewhere

**Comment by dfreilich** [comment]
@ipmb This is very interesting feedback. We typically perceived of it as two separate use cases, and one of the advantages of using `--publish` is that we can stream everything to a remote Docker registry, so the building process can be significantly faster (since it won't need to pull the run image). Having it both locally and publishing it would negate that improvement, and is counter to what the `--publish` flag means now, so I feel like that would be counterproductive. 

Can I ask why you would use the proposed workflow? If you're going to run tests, wouldn't you be running them against the image prior to pushing it to a remote repository? 

cc @ekcasey (with your feelings on supporting the non-daemon use case)

**Comment by ekcasey** [comment]
@dfreilich Reading between the lines here I suspect the real issue here is support for `--cache-image` when building a daemon image. From the lifecycle's perspective there is nothing wrong with using a cache image in a registry while exporting the app image to the docker daemon (the cache image itself can't be a deamon image, but the app image can). I would suggest that `pack` support this combination.

@ipmb Am I interpreting your needs correctly?

**Comment by ipmb** [comment]
@dfreilich Yes, testing then publishing would be a better workflow, but that would prevent me from using `--cache-image`

@ekcasey Yes, allowing `--cache-image` to work without `--publish` would solve my issue, assuming I can push the image manually after testing.

**Comment by dfreilich** [comment]
So publishing the cache image to the registry, but keeping the image local?

**Comment by ipmb** [comment]
> So publishing the cache image to the registry, but keeping the image local?

I hadn't considered that option, but it would work. Although it probably makes sense _not_ to publish the cache image if tests were to fail.

I was thinking of dumping both images (app & cache) into the local registry and then I can push them when tests pass.

**Comment by dfreilich** [comment]
@ipmb that's definitely the best current solution (and honestly, it's probably better than publishing a cache image, when the image it's connected to may not pass your tests). 

Would there still be an ask to make the workflow @ekcasey mentioned? (using/publishing a cache image in a registry, while having the app image be a local docker daemon image)

**Comment by ipmb** [comment]
> Would there still be an ask to make the workflow @ekcasey mentioned? (using/publishing a cache image in a registry, while having the app image be a local docker daemon image)

My workflow doesn't require that. Having `--publish` toggle publishing _all_ images feels intuitive to me.

**Comment by dfreilich** [comment]
What I felt as well. In that case, I'll close this issue. Thanks for bringing up the discussion (and for being an early user of the `--cache-image`!), and feel free to reach out if there's anything else we can do

**Comment by ipmb** [comment]
@dfreilich there's still an issue here that `--cache-image` *requires* the `--publish` flag currently. Would you like me to create a new issue for that or adjust this one accordingly.

**Comment by dfreilich** [comment]
I'm not sure I get which issue you're referencing, so to clarify:

1. Are you looking to use `--cache-image` to generate a local cache image (so to totally remove the `--publish` requirement)? That was discussed here (https://github.com/buildpacks/lifecycle/issues/484) in response to the initial PR (https://github.com/buildpacks/pack/pull/970), and @ekcasey noted that doing so apparently has abysmal performance when trying to push the cache image to the daemon, and recommended just using a volume cache instead. 
2. Are you trying to use an image generated by `--cache-image` in a (local) rebuild to speed up that 2nd build?

**Comment by ipmb** [comment]
My issue is that:

1. I'd like to use `--cache-image` to speed up future builds
2. Using `--cache-image` requires `--publish`
3. My workflow requires me to work with the image after building it (to run tests). When I use `--publish` I don't get an image locally. That means I have to download the image I just built from the registry. This seems inefficient.

I'm not sure of the most efficient solution here, but it seems like writing the image to disk then publishing it would be more efficient for my use case (no need to download anything from the registry). That workflow is incompatible with `--cache-image` which is something I'd like to use.

**Comment by dwillist** [comment]
Additional discussion here: https://buildpacks.slack.com/archives/C94UJCNV6/p1611603707053100.

Might want to think about how its possible to push multiple tags & have them go to local and remote location?

**Comment by juancoen** [comment]
Hi, I have a similar use case as the one described [here](https://github.com/buildpacks/pack/issues/1024#issuecomment-764946620). I want to publish the cache image but not the resulting app image, as I'm doing that later in my pipeline after running other non-buildpack tasks. If pushing the cache image to the local daemon is a problem, would it be possible to add a flag to only publish the cache image to the registry? Thanks!

**Comment by natalieparellano** [comment]
Just to add in case it's helpful - the lifecycle will not allow you to use a cache image in a docker daemon, for performance reasons (you're better off using a cache volume). Is the [earlier idea](https://github.com/buildpacks/pack/issues/1024#issuecomment-762901676) of allowing `--cache-image` regardless of `--publish` relevant to your use case?

**Comment by juancoen** [comment]
Yes, I think that will work for my use case, thanks!

**Comment by ipmb** [comment]
@natalieparellano What would using a cache volume look like. I see the `--volume` flag, but I'm not sure where I'd mount it.

Based on what you're saying, [for my use case](https://github.com/buildpacks/pack/issues/1024#issuecomment-764946620) it may be faster to have CI build the image in the docker daemon, run the necessary tests, upload the cache volume to S3 and push the image to the repository. Does that sound correct?

**Comment by natalieparellano** [comment]
@ipmb apologies for the slow reply. I think that makes sense - though I'd be curious to know why the cache volume needs to be uploaded. Is it being shared across builds?

`pack build` by default will create a cache volume to be used for all future builds of the same image. You can see this with the `-v` flag:

```
Using build cache volume pack-cache-library_bp-demo_latest-1413bf63c86b.build
```

Additionally there is some work underway to allow the volume to be specified: https://github.com/buildpacks/pack/issues/1077

**Comment by ipmb** [comment]
> though I'd be curious to know why the cache volume needs to be uploaded. Is it being shared across builds?

yeah, exactly. I'm using ephemeral build infrastructure (AWS Codebuild), so any resources I want to share across builds either need to get stored in a container registry or object store like S3 between builds. My assumption is that the cache volume is only useful if it exists for subsequent builds.

**Comment by jjbustamante** [comment]
I think this issue can be closed because:

- The following [RFC](https://github.com/buildpacks/rfcs/blob/main/text/0091-pack-cache-options.md) was created, approved and implemented by the issue #1077

**Comment by jjbustamante** [comment]
Closing this issue based on the implementations we already did on #1077
